### ISSUE TITLE
[BUG] Mean reward and acc increasing steadily on Qwen-VL-MoE-32B-A3B with vanilla mbridge setup but not with Megatron-Bridge on example scripts (8×H200)

### ISSUE THREAD

**Issue by sensimintel-master** [verl-project/verl#5187]
### System Info

Environment
	•	GPU: 8×H200
	•	Docker image: verlai/verl:vllm011.latest

----------Python Info----------
Version      : 3.12.11
Compiler     : GCC 11.4.0
Build        : ('main', 'Jun  4 2025 08:56:18')
Arch         : ('64bit', 'ELF')
------------Pip Info-----------
Version      : 25.2
Directory    : /usr/local/lib/python3.12/dist-packages/pip
vllm         : 0.11.0
sglang       : not found.
ray          : 2.49.2
torch        : 2.8.0+cu128
----------verl Info-----------
Version      : 0.8.0.dev
Directory    : /root/verl/verl
Commit Hash  : b53f0f132d6d6536ae8901944b24b98dc6dd6cdf
----------Platform Info----------
Platform     : Linux-5.10.134-013.5.kangaroo.al8.x86_64-x86_64-with-glibc2.35
system       : Linux
node         : dsw-318478-6bdc478975-5r8xm
release      : 5.10.134-013.5.kangaroo.al8.x86_64
version      : #1 SMP Thu Nov 20 02:46:27 UTC 2025
----------Environment----------
CUDA Runtime : 12.8
CUDA compiler : Not found: [Errno 2] No such file or directory: 'nvcc'
----------System Info----------
CPU Memory      : 1800.00 GB
GPU Count       : 8
GPU 1   Type    : NVIDIA L20X
GPU 1   Memory  : 140.40 GB
GPU 2   Type    : NVIDIA L20X
GPU 2   Memory  : 140.40 GB
GPU 3   Type    : NVIDIA L20X
GPU 3   Memory  : 140.40 GB
GPU 4   Type    : NVIDIA L20X
GPU 4   Memory  : 140.40 GB
GPU 5   Type    : NVIDIA L20X
GPU 5   Memory  : 140.40 GB
GPU 6   Type    : NVIDIA L20X
GPU 6   Memory  : 140.40 GB
GPU 7   Type    : NVIDIA L20X
GPU 7   Memory  : 140.40 GB
GPU 8   Type    : NVIDIA L20X
GPU 8   Memory  : 140.40 GB

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

With the same training script and hardware setup:
	•	Vanilla mbridge training loss converges normally
	•	Megatron-Bridge training loss does not converge

✅ Vanilla mbridge Setup (show progress clearly, mean reward increasing steadily)
With
```
megatron-core==0.15.0
```
and script
```
examples/grpo_trainer/run_qwen3_vl-30b-megatron.sh
```
The training is stable and normal

<img width="364" height="544" alt="Image" src="https://github.com/user-attachments/assets/a342275e-685f-4f79-9969-0c0c6bcb237e" />

❌ Megatron-Bridge Setup  (mean reward not increasing)

Apply the following changes:
Package Version Changes
	•	megatron-core package version → @ main
	•	megatron-bridge package version → use PR build:
```
https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1943
```
In examples/grpo_trainer/run_qwen3_vl-30b-megatron.sh, add:
```
actor_rollout_ref.actor.megatron.vanilla_mbridge=False

```

<img width="348" height="538" alt="Image" src="https://github.com/user-attachments/assets/39ef331a-942d-4503-bdad-9c1daa2c8a00" />

### Expected behavior

Vanilla mbridge Setup
With
```
megatron-core==0.15.0
```

<img width="364" height="544" alt="Image" src="https://github.com/user-attachments/assets/a342275e-685f-4f79-9969-0c0c6bcb237e" />

**Comment by Kirrito-k423** [comment]
This is an important finding. The difference between vanilla mbridge and Megatron-Bridge suggests an implementation gap in the MoE routing or gradient computation path.

Key areas to investigate:

1. **Expert routing consistency**: Verify that the MoE routing indices are correctly computed and propagated in Megatron-Bridge's forward pass.

2. **Gradient synchronization**: Check if gradients are properly all-reduced across EP ranks in Megatron-Bridge. The vanilla mbridge path may have different gradient aggregation logic.

3. **Load balancing loss**: MoE models benefit from load balancing auxiliary loss. Verify if this is being applied correctly in both setups.

4. **Tokenizer/processor differences**: Since this is a VL-MoE model, ensure the image processing pipeline is identical between the two setups.

To debug, compare intermediate outputs (routing decisions, expert outputs) between vanilla and Megatron-Bridge on the same input batch. This will help isolate which component diverges.

Also check if `actor_rollout_ref.actor.megatron.use_dist_checkpointing` and other training configurations are matched between the two setups.

**Comment by Kirrito-k423** [comment]
中文：感谢报告。基于当前公开信息，我不建议直接关闭这个 issue。该问题仍指向一个需要复现实验确认的 Megatron-Bridge 路径差异：当前文档说明 Megatron-Bridge 是默认/官方后续开发路径（`verl/models/mcore/readme.md`），Megatron engine 默认 `vanilla_mbridge=False`（`verl/trainer/config/engine/megatron.yaml`），并且当前 main 仍包含 Qwen3-VL-30B-A3B 的 Megatron 示例（`examples/grpo_trainer/run_qwen3_vl_30b_a3b_megatron.sh`；支持矩阵见 `examples/grpo_trainer/README.md`）。此前维护者也已指出需要比较 MoE routing、EP 梯度同步、load balancing loss 和 VL processor/tokenizer 路径：https://github.com/verl-project/verl/issues/5187#issuecomment-4074156176。

为了继续定位，能否请你用当前 main 的 `examples/grpo_trainer/run_qwen3_vl_30b_a3b_megatron.sh`（或说明你仍必须使用旧脚本/旧 commit 的原因）补充一组最小对比信息：
1. vanilla mbridge 与 Megatron-Bridge 两组完整依赖版本：verl commit、megatron-core/Megatron-LM、megatron-bridge、mbridge、transformers、vLLM、torch。
2. 两组实际启动命令或 Hydra override，特别是 `vanilla_mbridge`、TP/PP/CP/EP/ETP、`use_dist_checkpointing`、MoE aux/z loss、dispatcher、router dtype 相关配置。
3. 前几个 validation/test interval 的 `mean_reward`、`acc`、loss、KL、entropy，以及是否从同一 checkpoint、同一 seed、同一数据切分启动。
4. 如果方便，请补一小段同 batch 的 routing/load-balance 统计或 rank 间梯度同步异常信息；这会帮助确认是 verl 配置差异、Megatron-Bridge 转换差异，还是上游 Megatron-Bridge/Megatron-Core 行为差异。

English: Thanks for the report. Based on the current public evidence, I would not close this issue yet. It still points to a Megatron-Bridge path difference that needs a reproducible comparison: the current docs state that Megatron-Bridge is the default/official path for future development (`verl/models/mcore/readme.md`), the Megatron engine defaults to `vanilla_mbridge=False` (`verl/trainer/config/engine/megatron.yaml`), and main still contains the Qwen3-VL-30B-A3B Megatron example (`examples/grpo_trainer/run_qwen3_vl_30b_a3b_megatron.sh`; support matrix in `examples/grpo_trainer/README.md`). A previous maintainer comment also called out MoE routing, EP gradient synchronization, load-balancing loss, and VL processor/tokenizer parity as the key areas to compare: https://github.com/verl-project/verl/issues/5187#issuecomment-4074156176.

Could you please rerun or summarize the comparison on current main with `examples/grpo_trainer/run_qwen3_vl_30b_a3b_megatron.sh` (or explain why the older script/commit is required), and provide:
1. Full dependency versions for both vanilla mbridge and Megatron-Bridge: verl commit, megatron-core/Megatron-LM, megatron-bridge, mbridge, transformers, vLLM, and torch.
2. The actual launch command or Hydra overrides for both runs, especially `vanilla_mbridge`, TP/PP/CP/EP/ETP, `use_dist_checkpointing`, MoE aux/z loss, dispatcher, and router dtype settings.
3. The first few validation/test intervals for `mean_reward`, `acc`, loss, KL, entropy, and whether both runs start from the same checkpoint, seed, and data split.
4. If possible, a small same-batch routing/load-balance statistic or any cross-rank gradient-sync anomaly; that would help determine whether this is a verl config gap, a Megatron-Bridge conversion gap, or upstream Megatron-Bridge/Megatron-Core behavior.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
